src/fairseq2/metrics.py

```diff
@@ -158,24 +158,218 @@ def sync_and_compute_metrics(*bags: MetricBag) -> Optional[Dict[str, Any]]:
     return values
 
 
-def format_metric_values(values: Dict[str, Any], *bags: MetricBag) -> str:
-    """Format metric ``values`` to print to stdout or stderr."""
-    fmt_values = []
+def format_as_int(value: Any, *, postfix: Optional[str] = None) -> str:
+    """Format metric ``value`` as integer."""
+    i = int(value)
 
-    for name, value in values.items():
-        fmt_value: Optional[str]
+    s = "<1" if i == 0 and isinstance(value, float) else f"{i:,}"
 
-        if name == "wall_time":
-            fmt_value = f"{int(value)}s"
-        else:
-            for bag in bags:
-                fmt_value = bag.format_metric_value(name, value)
-                if fmt_value is not None:
-                    break
+    if postfix:
+        s += postfix
+
+    return s
+
+
+format_as_seconds = partial(format_as_int, postfix="s")
+"""Format metric ``value`` as duration in seconds."""
+
+
+def format_as_float(
+    value: Any, *, decimal: Optional[int] = None, postfix: Optional[str] = None
+) -> str:
+    """Format metric ``value`` as float."""
+    if decimal:
+        s = f"{float(value):,.{decimal}f}"
+    else:
+        s = f"{float(value):,}"
+
+    if postfix:
+        s += postfix
+
+    return s
+
+
+format_as_loss = partial(format_as_float, decimal=3)
+"""Format metric ``value`` as training loss."""
+
+
+_metric_formatters: Dict[str, Tuple[str, Callable[[Any], str]]] = {
+    # fmt: off
+    "batch_size":          ("Batch Size",                format_as_int),
+    "elapsed_time":        ("Elapsed Time",              format_as_seconds),
+    "elements_per_batch":  ("Elements per Batch",        format_as_int),
+    "elements_per_second": ("Elements per Second",       format_as_int),
```

Comment:
this can be a float right ? e.g. if we process 1 element per 400 ms ?

Reply:
These formatters are meant to be used to report progress via stdout/stderr. For values less that 1.0, we just show `<1`. For instance, see the elapsed time in this log output:

```
Valid:eng-yor Metrics (step 200) - Loss: 10.645 | Entropy Loss: 15.358 | Batch Size: 248 | Elements per Batch: 11,304 | Elements per Second: 46,781 | Number of Source Elements: 33,447 | Number of Target Elements: 45,217 | Elapsed Time: <1s
```

Reply:
Thank you for the clarification!

Reply:
No problem!